ledger/store/src/program/finalize.rs

```diff
@@ -553,196 +406,134 @@ pub trait FinalizeStorage<N: Network>: 'static + Clone + Send + Sync {
+            Some(value) => Ok(Some(cow_to_cloned!(value))),
             None => Ok(None),
         }
     }
 
     /// Returns the speculative value for the given `program ID`, `mapping name`, and `key`.
     fn get_value_speculative(
         &self,
-        program_id: &ProgramID<N>,
-        mapping_name: &Identifier<N>,
+        program_id: ProgramID<N>,
+        mapping_name: Identifier<N>,
         key: &Plaintext<N>,
     ) -> Result<Option<Value<N>>> {
-        // Retrieve the key ID.
-        match self.get_key_id_speculative(program_id, mapping_name, key)? {
-            // Retrieve the value.
-            Some(key_id) => self.get_value_from_key_id_speculative(&key_id),
-            None => Ok(None),
-        }
-    }
-
-    /// Returns the confirmed value for the given `key ID`.
-    fn get_value_from_key_id_confirmed(&self, key_id: &Field<N>) -> Result<Option<Value<N>>> {
-        match self.value_map().get_confirmed(key_id)? {
+        match self.key_value_id_map().get_value_speculative(&(program_id, mapping_name), key)? {
             Some(value) => Ok(Some(cow_to_cloned!(value))),
             None => Ok(None),
         }
     }
 
-    /// Returns the speculative value for the given `key ID`.
-    fn get_value_from_key_id_speculative(&self, key_id: &Field<N>) -> Result<Option<Value<N>>> {
-        match self.value_map().get_speculative(key_id)? {
-            Some(value) => Ok(Some(cow_to_cloned!(value))),
-            None => Ok(None),
-        }
+    /// Returns the confirmed checksum of the finalize storage.
+    fn get_checksum_confirmed(&self) -> Result<Field<N>> {
+        // Compute all mapping checksums.
+        let preimage: std::collections::BTreeMap<_, _> = self
+            .key_value_id_map()
+            .iter_confirmed()
+            .map(|(m, k, v)| {
+                let m = cow_to_copied!(m);
+                let k = cow_to_cloned!(k);
+                let v = cow_to_cloned!(v);
+
+                let mut preimage = Vec::new();
+                m.write_bits_le(&mut preimage);
+                false.write_bits_le(&mut preimage); // Separator.
+                k.write_bits_le(&mut preimage);
+                false.write_bits_le(&mut preimage); // Separator.
+
+                // Compute the mapping checksum as `Hash( m || k )`.
+                let mapping_checksum = N::hash_bhp1024(&preimage)?;
+
+                v.write_bits_le(&mut preimage);
+                false.write_bits_le(&mut preimage); // Separator.
+
+                // Compute the entry checksum as `Hash( m || k || v )`.
+                let entry_checksum = N::hash_bhp1024(&preimage)?;
+                // Return the mapping checksum and entry checksum.
+                Ok::<_, Error>((mapping_checksum, entry_checksum.to_bits_le()))
+            })
+            .try_collect()?;
+        // Compute the checksum as `Hash( all mapping checksums )`.
+        N::hash_bhp1024(&preimage.into_values().flatten().collect::<Vec<_>>())
```

Comment:
since we're only interested in the values, do we need to calculate the `mapping_checksum` values? or do we expect there to be some duplicates that get overwritten?

Reply:
its necessary to sort the values.